File: api-testing-portfolio/playwright-api/clients/booking.client.ts
import type { APIRequestContext, APIResponse } from '@playwright/test';
import type { BookingPayload } from '../data/booking.factory';

/**
 * Restful Booker authenticates write operations with a token passed in a
 * Cookie header. The client accepts the token per call so the same instance
 * can exercise authenticated, unauthenticated and invalid-token scenarios.
 */
export class BookingClient {
  constructor(private readonly request: APIRequestContext) {}

  private authHeaders(token?: string): Record<string, string> {
    return token ? { Cookie: `token=${token}` } : {};
  }

  async create(booking: BookingPayload): Promise<APIResponse> {
    return this.request.post('/booking', { data: booking });
  }

  /** Sends a raw (potentially malformed) body instead of a JSON object. */
  async createRaw(rawBody: string): Promise<APIResponse> {
    return this.request.post('/booking', {
      headers: { 'Content-Type': 'application/json' },
      data: rawBody,
    });
  }

  /** Allows arbitrary payload shapes for negative tests (missing/typed-wrong fields). */
  async createPartial(payload: Record<string, unknown>): Promise<APIResponse> {
    return this.request.post('/booking', { data: payload });
  }

  async getById(id: number): Promise<APIResponse> {
    return this.request.get(`/booking/${id}`);
  }

  async getIds(filters: Record<string, string> = {}): Promise<APIResponse> {
    return this.request.get('/booking', { params: filters });
  }

  async update(id: number, booking: BookingPayload, token?: string): Promise<APIResponse> {
    return this.request.put(`/booking/${id}`, {
      headers: this.authHeaders(token),
      data: booking,
    });
  }

  async partialUpdate(
    id: number,
    partial: Partial<BookingPayload>,
    token?: string,
  ): Promise<APIResponse> {
    return this.request.patch(`/booking/${id}`, {
      headers: this.authHeaders(token),
      data: partial,
    });
  }

  async delete(id: number, token?: string): Promise<APIResponse> {
    return this.request.delete(`/booking/${id}`, { headers: this.authHeaders(token) });
  }
}

File: api-tests/src/test/java/booker/clients/BookingClient.java
package booker.clients;

import booker.models.Booking;
import booker.utils.Config;
import io.restassured.RestAssured;
import io.restassured.response.Response;
import io.restassured.specification.RequestSpecification;

import java.util.Map;

public class BookingClient {

    public BookingClient() {
        RestAssured.baseURI = Config.baseUrl();
    }

    private RequestSpecification base() {
        return RestAssured.given()
                .header("User-Agent", "Mozilla/5.0")
                .header("Connection", "close")
                .contentType("application/json")
                .accept("application/json");
    }

    public Response create(Booking booking) {
        return base().body(booking).when().post("/booking");
    }

    public Response createRaw(String rawJson) {
        return base().body(rawJson).when().post("/booking");
    }

    public Response createPartial(Map<String, Object> partial) {
        return base().body(partial).when().post("/booking");
    }

    public Response getById(int id) {
        return base().when().get("/booking/" + id);
    }

    public Response getIds(Map<String, ?> filters) {
        return base()
                .queryParams(filters == null ? Map.of() : filters)
                .when()
                .get("/booking");
    }

    public Response update(int id, Booking booking, String token) {
        var req = base().body(booking);
        if (token != null) {
            req.header("Cookie", "token=" + token);
        }
        return req.when().put("/booking/" + id);
    }

    public Response partialUpdate(int id, Map<String, Object> partial, String token) {
        var req = base().body(partial);
        if (token != null) {
            req.header("Cookie", "token=" + token);
        }
        return req.when().patch("/booking/" + id);
    }

    public Response delete(int id, String token) {
        var req = base();
        if (token != null) {
            req.header("Cookie", "token=" + token);
        }
        return req.when().delete("/booking/" + id);
    }
}
